dimagi/commcare-hq
# Trigger Geopoint ES Index on Geospatial Feature Flag Enable

## Product Description

When the Geospatial feature flag is enabled, the domain will receive the following alert on all Geospatial pages which show the progress of indexing the domain's location properties:
![Screenshot from 2024-09-16 11-36-30](https://github.com/user-attachments/assets/e7c25745-8886-428a-abf3-d6c39847bf44)

If, however, the domain's case count is over the limit then the following error alert will be displayed instead:
![Screenshot from 2024-09-16 11-35-32](https://github.com/user-attachments/assets/6fe4aa63-3735-4b3e-b300-505588d76c84)

## Technical Summary

Link to ticket [here](https://dimagi.atlassian.net/browse/SC-3864).
Link to tech spec [here](https://docs.google.com/document/d/1bLg-pTzLrd62ExNAb9Fdh6zCJzJ11gWUZTtlz1L7HY8/edit).

A new Celery task has been added which will be automatically started when the `GEOSPATIAL` feature flag is enabled for a domain. This new task essentially runs through the same logic as the `index_geolocation_case_properties` management command, where ES docs for a domain have their location case properties indexed to add a geopoint property. This is done so that these ES docs can be used within the context of the Geospatial feature.

The `CeleryTaskTracker` model has been expanded to support the ability of storing a message in Redis, as well as having an error status. These are used to correctly show either the task progress or applicable error message to the user on the front-end.

## Feature Flag

`GEOSPATIAL`

## Safety Assurance

### Safety story

- Local testing
- QA done

This change will only affect domains that enable the Geospatial feature flag. Additionally, an ES doc limit has been put in place so that the Celery task only starts indexing if the domain has cases fewer than the specified limit. The Celery task itself has also been set to run as a serial task, and so it is expected that no more than one of these tasks will be active at a time in the environment.

### Automated test coverage


Automated tests exist for the new Celery task that has been added. Additionally, the tests for the `CeleryTaskHelper` class have been extended to account for the new changes there.

### QA Plan


QA has been completed. Ticket is available [here](https://dimagi.atlassian.net/browse/QA-7112).

### Rollback instructions




### Labels & Review
